##############################
####Arvis Sulovari    ########
#####STR/VNTR genotyping######
##########01/18/2018##########
##############################

import os

if config == {}:
    configfile: "strvntr_pipeline.json"

HAPLOTYPE_FILE = config["samples"]
DATA_DIR = config["fasta_dir"]
RANGES_FILE = config["ranges"]

#Read in each haplotype names (11 total)
HAPLOTYPES = []
with open(HAPLOTYPE_FILE) as f:
    HAPLOTYPES = f.read().splitlines()
    print(HAPLOTYPES)

RANGES = []
with open(RANGES_FILE) as f:
    RANGES = f.read().splitlines()
    print(RANGES)


#Expand name of haplotype files
rule all:
    input: expand("{haplotype}.{range}.tab", haplotype = HAPLOTYPES, range = RANGES)
    params: sge_opts="-l mfree=4G -N rule_all" 


rule PureRep_finder:
    #input: "{haplotype}","{range}"
    output: "{haplotype}.{range}.tab"
    log: "log/{haplotype}.{range}.PureRep.log"
    params: sge_opts="-l mfree=2G -N pure_rep" 
    shell: """
    module purge
    module load modules modules-init modules-gs mpc/0.8.2 mpfr/3.1.0 gmp/5.0.2 gcc/4.9.1 R/3.4.0

    Rscript str_vntr_genotyper.R {wildcards.haplotype} {wildcards.range}
    
    cp {wildcards.haplotype}.{wildcards.range}.tab {wildcards.haplotype}/
    """



